Allow running changed packages on target branches

Josh-ES opened the pull request:
I'm not sure whether this fits with where the maintainers see Circletron going, but I thought I'd open a PR to see what you all think.

Right now, every change to a target branch triggers all jobs across all packages. My feeling is that as a user of the orb, I should be able to set a configuration flag to only run jobs against changed packages.

insidewhy commented on src/git.ts:
```diff
@@ -75,3 +75,13 @@ export const getBranchpointCommit = async (mainBranchRegex: RegExp): Promise<str
+export const getLastCommitOnBranch = async (): Promise<string> => {
+  const stdout = await spawnGetStdout('git', ['log', '-n', '2', '--pretty=format:%h'])
+
+  if (stdout.trim().split('\n').length > 1) {
```
```suggestion
  if (stdout.trim().split('\n', 2).length > 1) {
```

insidewhy commented on src/git.ts:
```diff
@@ -75,3 +75,13 @@ export const getBranchpointCommit = async (mainBranchRegex: RegExp): Promise<str
+  const stdout = await spawnGetStdout('git', ['log', '-n', '2', '--pretty=format:%h'])
+
+  if (stdout.trim().split('\n').length > 1) {
+    return stdout.trim().split('\n')[1]
```
I'd probably cache the same call above in a variable and return it here.

insidewhy commented:
> I'm not sure whether this fits with where the maintainers see Circletron going, but I thought I'd open a PR to see what you all think.
> 
> Right now, every change to a target branch triggers all jobs across all packages. My feeling is that as a user of the orb, I should be able to set a configuration flag to only run jobs against changed packages.

The problem here is this solution gets the "commit before last", but unless the merge to the target branch was only one commit, it will miss changes from the branch that was merged. Definitely for merges via rebase, and I think maybe this command also reaches into the commits of the merge rather than getting the commit before the branch.

Josh-ES commented:
> The problem here is this solution gets the "commit before last", but unless the merge to the target branch was only one commit, it will miss changes from the branch that was merged. Definitely for merges via rebase, and I think maybe this command also reaches into the commits of the merge rather than getting the commit before the branch.

Very good point. I've replaced the git command in `getLastCommitOnBranch` with `git log HEAD@{1} -1`, which should cover merge via rebase. I also added some error handling so that if no previous commit is found, all pipelines are run - I believe this is necessary to handle the first commit on a target branch.

insidewhy commented:
@Josh-ES `git log HEAD@{1} -1` doesn't work for me, at least not on any test I ran on my local git repository. Is it something specific to do with how circleci performs the checkout?

Josh-ES commented:
@insidewhy I was thinking about this a bit more deeply, and I've realised that we don't just want to get the last commit before the merge. We probably want to get the last commit before the merge that built successfully in Circle.

I've updated this implementation to use the Circle API (disappointingly, this is a little convoluted with Circle CI's API, requiring multiple calls) to try to find the git revision corresponding to the last successful build on the current branch. I think this should satisfy both of our concerns about finding the commit before a merge and excluding commits where the build failed.

Let me know your thoughts!

Josh-ES commented on src/circle.ts:
```diff
@@ -0,0 +1,148 @@
+    // given the build URL, which is of the form https://circleci.com/api/v2/project/{project_slug}/{build_number}
+    // we can extract the project slug to call the API
+    const buildUrl = requireEnv('CIRCLE_BUILD_URL')
+    const projectSlug: string | undefined = /circleci\.com\/(.*\/.*\/.*)\//.exec(buildUrl)?.[1]
```
Yes sorry for the contradiction, the comment is incorrect rather than the code. I'll update.

Josh-ES commented on src/circle.ts:
```diff
@@ -0,0 +1,148 @@
+}
+
+interface CircleWorkflowItem {
+  // The unique ID of the workflow
```
I was blindly copying notes down from the Circle API docs to be honest, but these are glaringly obvious. I'll remove.

Josh-ES commented on src/circle.ts:
```diff
@@ -0,0 +1,148 @@
+ */
+export async function getLastSuccessfulBuildRevisionOnBranch(
+  branch: string,
+): Promise<string | null> {
```
I've always felt that `null` should be used in a case like this where we're explicitly saying that there is no appropriate revision to return. `undefined` is more that something hasn't been defined yet. I've made the change, but I'm interested to hear more about your opinion here

insidewhy replied:
I hate that JavaScript has `null` and `undefined`, it's such an unnecessary complication. I just try only use one, and when sticking to one, it's usually easier to stick to `undefined` or you can end up with a lot of `?? null`. Of course this can get frustrating for APIs that return `null` 😭

Josh-ES replied:
Ah I see! That's interesting to hear. Not something I've heard before, but that sounds reasonable and I'll stick to `undefined`

Josh-ES commented on src/circle.ts:
```diff
@@ -0,0 +1,148 @@
+            )
+
+            return (workflowData as CircleWorkflows).items.every(
+              (item) => item.status === CircleWorkflowStatus.Success,
```
I've updated this to also include workflows in an 'on hold' state. I feel this in effect assumes that whatever follows the hold job will succeed, while if something after a hold job fails, I'd expect to re-run that pipeline

Josh-ES commented on src/circle.ts:
```diff
@@ -0,0 +1,148 @@
+    if (projectSlug) {
+      // Call the API for pipelines, which tell us nothing about whether all the workflows within them were
+      // successful or not
+      const { data: pipelineData } = await axios.get(
```
There are 20 pipelines per page, and of course I don't currently plan to go beyond the first page of results. I've tested this change in a situation where all 20 jobs had failing workflows, and the trigger job took ~2s according to Circle's dashboard instead of ~1s, so I felt that was reasonable

insidewhy commented:
@Josh-ES Thanks for this work, it's really interesting. I'm gonna get a few more of my colleagues to review it and then merge it if everything looks good. We might try it out on our own projects to shave off some time.

insidewhy commented:
@Josh-ES Also, do you want me to push an orb release with the `dev:next` tag so that you can test this out?

sxwei123 commented on readme.md:
````diff
@@ -63,3 +63,18 @@ This will cause jobs within `project1` to run when changes are detected in eithe
+targetBranches: ^(release/|main$|master$|develop$)
+```
+
+On target branches, Circletron can be configured to only run workflows in the packages that have changed since the last successful build on that branch. This feature interacts with the Circle API v2, so requires an access token to be provided, via the `CIRCLE_TOKEN` environment variable. You can turn this on using `runOnlyChangedOnTargetBranches`:
````
```suggestion
On target branches, Circletron can be configured to only run workflows in the packages that have changed since the last successful build on that branch. This feature interacts with the Circle API v2, so requires an access token to be provided, via the `CIRCLE_TOKEN` environment variable. Turn on this feature by enabling `runOnlyChangedOnTargetBranches`:
```

insidewhy replied:
Ah yeah @sxwei123 has a good point here, the documentation is currently in impersonal mood, it'd be nice to stay consistent on that. Impersonal mood usually ends up terser and I think it sounds a little more professional.

sxwei123 commented on src/circle.ts:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    if (projectSlug) {
+      // Call the API for pipelines, which tell us nothing about whether all the workflows within them were
```
```suggestion
      // Call the API for pipelines, which tell nothing but whether all the workflows within them were
```

insidewhy commented on readme.md:
```diff
@@ -63,3 +63,18 @@ This will cause jobs within `project1` to run when changes are detected in eithe
+
+## Configuration Properties
+
+Circletron configuration can be provided in a `.circle/circletron.yml` file. You can specify target branches using `targetBranches`:
```
It might be better to explain the `targetBranches` configuration after explaining what target branches are rather than before?

insidewhy replied:
Also there is already some documentation on line 31-38 that explains what this configuration is, I agree it's nice to group the configuration properties into a section, but would want to avoid duplicating information.

sxwei123 commented on src/circle.ts:
```diff
@@ -0,0 +1,130 @@
+    if (projectSlug) {
+      // Call the API for pipelines, which tell us nothing about whether all the workflows within them were
+      // successful or not
+      const { data: pipelineData } = await axios.get(
+        `${CIRCLE_API_URL}/project/${projectSlug}/pipeline`,
+        {
+          headers,
+          params: {
+            branch,
+          },
+        },
+      )
+
+      // For each pipeline, fetch the workflows and find the first one where all the workflows have a
+      // 'success' or 'on hold' status. We assume that 'on hold' pipelines would have succeeded were
+      // they to be approved
+      const lastSuccessfulBuild = await find(
+        (pipelineData as CirclePipelines).items,
```
```suggestion
      const { data: pipelineData } = await axios.get<CirclePipelines>(
        `${CIRCLE_API_URL}/project/${projectSlug}/pipeline`,
        {
          headers,
          params: {
            branch,
          },
        },
      )

      // For each pipeline, fetch the workflows and find the first one where all the workflows have a
      // 'success' or 'on hold' status. We assume that 'on hold' pipelines would have succeeded were
      // they to be approved
      const lastSuccessfulBuild = await find(
        pipelineData.items,
```

sxwei123 commented on src/index.ts:
```diff
@@ -195,14 +203,18 @@ async function buildConfiguration(
 
   return {
+    runOnlyChangedOnTargetBranches:
+      rawConfig.runOnlyChangedOnTargetBranches !== undefined
+        ? rawConfig.runOnlyChangedOnTargetBranches
+        : DEFAULT_RUN_ONLY_CHANGED_ON_TARGET_BRANCHES,
```
```suggestion
      rawConfig.runOnlyChangedOnTargetBranches ?? DEFAULT_RUN_ONLY_CHANGED_ON_TARGET_BRANCHES,
```

insidewhy commented:
@Josh-ES I've merged this now and released circletron 3.0.3 with the changes. Please try it out and let me know if it's working.

Josh-ES commented:
@insidewhy many thanks for merging and releasing, I appreciate it and will give it a test when I next get the chance in a few hours. Would you like me to circle back and address some of the above comments that appear to be unresolved?

insidewhy commented:
@Josh-ES I think we fixed them all.

Josh-ES commented:
@insidewhy Great! S Sounds good. I've tested this, both prior to the change with my own published version of the orb and with the new version and everything's working as intended.
